host/crypto/bcrypt/cert.c

```diff
@@ -0,0 +1,1291 @@
+        else
+            OE_RAISE_MSG(
+                OE_CRYPTO_ERROR,
+                "CertGetCRLFromStore failed, err=%#x\n",
+                error);
+    }
+
+    /* Given an existing validation, it's also possible (albeit irrgular) for a
+     * higher CA to revoke the cert still. */
+    if (cert_issuers_count > 1)
+    {
+        oe_result_t parent_result = _bcrypt_check_revocation(
+            cert, &cert_issuers[1], cert_issuers_count - 1, cert_store);
+
+        /* The parent issuer not having an applicable CRL for the cert is not
+         * necessarily an error, since is_validated may already be true. */
+        if (parent_result == OE_OK)
+            is_validated = true;
+        else if (parent_result != OE_VERIFY_CRL_MISSING)
+            OE_RAISE(parent_result);
+    }
+
+    /* Check if any of the CRLs were applicable to the cert */
+    if (!is_validated)
+        OE_RAISE_MSG(
+            OE_VERIFY_CRL_MISSING,
+            "No CRLs found were applicable to the cert.\n",
+            NULL);
+
+    result = OE_OK;
+
+done:
+    if (crl)
+        CertFreeCRLContext(crl);
+
+    return result;
+}
+
+static oe_result_t _bcrypt_verify_cert_chain(
+    HCERTSTORE cert_store,
+    PCCERT_CHAIN_CONTEXT cert_chain)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    CERT_CHAIN_POLICY_STATUS policy_status = {
+        .cbSize = sizeof(CERT_CHAIN_POLICY_STATUS)};
+
+    if (!_find_root_cert(cert_chain))
+        OE_RAISE_MSG(OE_VERIFY_FAILED, "No root certificate found\n", NULL);
+
+    /* NOTE: This only validates the cert chain structure for basic
+     * constraints such as path length. Cert chains generated via
+     * CertGetCertificateChain should already be valid in that regard.
+     *
+     * Ideally, this should try to run CERT_CHAIN_POLICY_AUTHENTICODE,
+     * which will also check cert_chain->TrustStatus for signature and
+     * revocation errors, but given the current parameterization, it is
+     * flagging CERT_TRUST_IS_NOT_SIGNATURE_VALID which seems to be flowing
+     * from CERT_TRUST_IS_UNTRUSTED_ROOT, which is expected given that the
+     * attestation cert trust is independent of the Windows root trust
+     */
+    if (!CertVerifyCertificateChainPolicy(
+            CERT_CHAIN_POLICY_BASIC_CONSTRAINTS,
+            cert_chain,
+            (PCERT_CHAIN_POLICY_PARA)&_OE_DEFAULT_CERT_CHAIN_POLICY,
+            &policy_status))
+    {
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR,
+            "CertVerifyCertificateChainPolicy could not check policy\n",
+            NULL);
+    }
+
+    /* Success on CertVerifyCertificateChainPolicy does not indicate
+     * successful validation, that's reflected in the policy_status */
+    if (policy_status.dwError != ERROR_SUCCESS)
+        OE_RAISE_MSG(
+            OE_VERIFY_FAILED,
+            "CertVerifyCertificateChainPolicy failed, err=%#x\n",
+            policy_status.dwError);
+
+    /* Explicitly walk the chain and validate the signatures */
```

Comment:
Not completely, my observations from the disassembly walk without debug symbols are recorded in the comment on prior to the `CertVerifyCertificateChainPolicy()` invocation on line 427, but it's not definitive. 

It looked like `CERT_TRUST_IS_UNTRUSTED_ROOT` was hit and causing a lot of other checks to be elided and immediately invalidate the signature, since we don't register the root cert trust in Windows root cert store explicitly but it's not clear if there's some way around that.